### PR TITLE
Http error callback handler

### PR DESCRIPTION
**Pull request by Rom4eg**
Added the ability to create callbacks for http errors in http-bind. There is no JQuery or other third party libraries.

Example:

myScript.js:

```JavaScript
function Error500(err_code){
    console.log(this); // Strophe.Connection object
    console.log('Error ' + err_code);
}

$(document).ready(function(){
    var conn = Strophe.connect('http://example.com/http-bind');
    conn.addHttpErrHndl(500, Error500);
    conn.connect('user_name@incorrect_jabber_host', 'secret', onConnect);
})
```

p.s Sorry for my english.

### CODE REVIEW

**Comment by Gordin** [comment]
If you want this merged, please make the changes in the source files instead of the compiled strophe.js file. Also, If I'm not mistaken, this looks like it only affects bosh and not websockets? If so, please make sure most of your changes are in src/bosh.js.
Currently you are modifying Strophe.Connection for the error handler, but the handler and most of the code should be be in Strophe.Bosh if it only affects Bosh. Look at how .attach is handled (function in Strophe.Connection that calls .attach on the ._proto function that has the actual logic)

If you've done this also include the compiled strophe.js file in the pull request

**Comment by jcbrand** [comment]
Additionally, `addHttpErrHndlr` doesn't conform to the naming convention used in Strophe.js.

The other similar methods are called `addHandler` and `addTimedHandler`. So please rename your method to `addHTTPErrorHandler`.

**Comment by Rom4eg** [comment]
Everything is ready.

1. Error handlers(```httpErrorHandler```) stored in Connection object, like other handlers.
2. Method ```addHTTPErrorHandler``` stored in Connection object
3. Action that trigger handler stored in Bosh object.

What was done after changes.
1. make check

```
Took 178ms to run 70 tests. 70 passed, 0 failed.
```
2. make 
```
rom4eg@home:~/Strophe/strophejs$ make
make strophe.js
make[1]: Вход в каталог `/home/rparanichev/Strophe/strophejs'
make stamp-bower
make[2]: Вход в каталог `/home/rparanichev/Strophe/strophejs'
make[2]: `stamp-bower' не требует обновления.
make[2]: Выход из каталога `/home/rparanichev/Strophe/strophejs'
Building strophe.js ...
./node_modules/.bin/grunt concat
Running "concat:dist" (concat) task
File "strophe.js" created.

Done, without errors.

make[1]: Выход из каталога `/home/rparanichev/Strophe/strophejs'
Building strophe.min.js ...
./node_modules/.bin/grunt min
Running "concat:dist" (concat) task
File "strophe.js" created.

Running "uglify:dist" (uglify) task
File "strophe.min.js" created.

Done, without errors.
```

File: src/core.js [hunk 1]
```diff
@@ -1653,6 +1655,34 @@ Strophe.Connection.prototype = {
         }
     },
 
+    /** Function: addHTTPErrorHandler
+     *  Add callback to HTTP error
+     *
+     *  The ability to create a callback function for HTTP errors.
+     *
+     *  This handlers will fire only with http-bind service only.
```
Comment (jcbrand) [inline src/core.js]: Grammar: `These handlers will fire only in the case of BOSH and the http-bind service.`

File: src/core.js [hunk 2]
```diff
@@ -1527,6 +1527,8 @@ Strophe.Connection = function (service, options)
     this.removeHandlers = [];
     this.addTimeds = [];
     this.addHandlers = [];
+    //HttpError handlers
+    this.httpErrorHandler = {};
```
Comment (jcbrand) [inline src/core.js]: This object contains potentially multiple handlers, so please name it `httpErrorHandlers` (like the ones above it).

File: src/bosh.js [hunk 3]
```diff
@@ -391,6 +391,11 @@ Strophe.Bosh.prototype = {
                      ", number of errors: " + this.errors);
         if (this.errors > 4) {
             this._conn._onDisconnectTimeout();
+        } else {
+          var err_callback = this._conn.httpErrorHandler[reqStatus];
+          if(err_callback){
+            err_callback.apply(this, [reqStatus,]);
+          }
         }
```
Comment (jcbrand) [inline src/bosh.js]: Please stick to the style used in Strophe. We indent 4 spaces, not 2 and there should be a space between `if` and the brackets `(err_callback)`.

**Comment by jcbrand** [comment]
@Rom4eg Looks good to me. There are just 3 small style and language issues.

**Comment by Rom4eg** [comment]
Done.

**Comment by jcbrand** [comment]
Thanks! @Gordin if you're happy I'll merge.

**Comment by Gordin** [comment]
Hm, the code looks OK, but I was wondering if it would be a good idea to turn this into an "addProtocolErrorHandler" that would also work for WebSockets. That handler could be used in the  socket.onerror callback that currently doesn't have a handler and just disconnects.
@jcbrand Thoughts?
(I'd merge this and do the implementation myself)

**Comment by Rom4eg** [comment]
Don't merge!!!!

**Comment by Rom4eg** [comment]
One second, i find an issue

**Comment by Rom4eg** [comment]
Now it's ok.

Sorry.

**Comment by jcbrand** [comment]
@Gordin: I think that's a great idea.

**Comment by kabilan-baskaran** [comment]
@jcbrand : Thank you very much .It makes sense.
 However,it only solves our problem partially .Since it returns only the status code not the complete  stanza.So the end user don't know for which stanza the success/error happened . It's good to have the complete stanza in return function as another parameter.

**Comment by jcbrand** [comment]
Yes, I agree that the callback should be passed the stanza.

**Comment by kabilan-baskaran** [comment]
@jcbrand :+1: Thanks .Waiting for your next pull.

**Comment by jcbrand** [comment]
@kabilan-baskaran I'm not working on this ticket and am not going to make a pull request, it's @Gordin who said that he wants to expand on this pull request.

**Comment by kabilan-baskaran** [comment]
Ok.No problem.

**Comment by Dashue** [comment]
Bump?
I'd like a way to receive know about the http errors as well. Where did this get left off?

**Comment by Dashue** [comment]
Think I found a way around it through
xmppConnection.connection.addHandler(onError, null, null, 'error')

**Comment by jcbrand** [comment]
I've refactored the code in the pull request to make it protocol agnostic (although websocket implementation is not done yet).

The handler is now added via `addProtocolErrorHandler` and the first parameter is the protocol (either `'HTTP`' or `'websocket'`).

**Comment by Dashue** [comment]
This was definitely the biggest issue we found ourselves when working with this. Good to see progress on it!

**Comment by jcbrand** [comment]
@Dashue I'm planning on making another release soon, it would be great if you could first confirm that these changes work for you.

**Comment by Dashue** [comment]
Unfortunately not working with this anymore and won't be of any help.

**Pull request merged**